### PR TITLE
(FM-5119) Automation test for IM on AIX

### PR DESCRIPTION
**Pull request by phongdly**

### CODE REVIEW

File: tests/integration/pre-suite/01_websphere_module_install.rb [hunk 1]
```diff
@@ -1,29 +1,22 @@
...
+    on(agent, "/opt/puppetlabs/puppet/bin/puppet module install dsestero-download_uncompress")
   end
 end
 
-# When the module puppetlabs-ibm_installation_manager is ready for testing. The below lines of code will be uncommented.
+step 'Install ibm_installation_manager Module'
+proj_root = File.expand_path(File.join(File.dirname(__FILE__), '../../../'))
+staging = { :module_name => 'puppetlabs-ibm_installation_manager' }
+local = { :module_name => 'ibm_installation_manager', :source => proj_root,
+          :target_module_path => '/etc/puppetlabs/code/environments/production/modules'}
```
Comment (cowofevil) [inline tests/integration/pre-suite/01_websphere_module_install.rb]: Inconsistent spacing around braces.

File: tests/test_run_scripts/integration_tests_aix.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+SCRIPT_PATH=$(pwd)
+BASENAME_CMD="basename ${SCRIPT_PATH}"
+SCRIPT_BASE_PATH=`eval ${BASENAME_CMD}`
+
+if [ $SCRIPT_BASE_PATH = "test_run_scripts" ]; then
+  cd ../../
+fi
+
+# Work-around for RE-5005
+export SSL_CERT_FILE=/usr/local/etc/openssl/cert.pem
```
Comment (cowofevil) [inline tests/test_run_scripts/integration_tests_aix.sh]: This breaks on my system. I would suggest leaving it out of your run scripts.

File: tests/integration/pre-suite/01_websphere_module_install.rb [hunk 3]
```diff
@@ -1,29 +1,22 @@
...
+step 'install IBM_Installation_Manager dependencies:'
+on(master, puppet('module install puppet-archive'))
 on(master, puppet('module install puppetlabs-stdlib'))
 on(master, puppet('module install puppetlabs-concat'))
 
 step 'Install dsestero/download_uncompress module on the agent'
 confine_block(:except, :roles => %w{master dashboard database}) do
   agents.each do |agent|
-    on(agent, puppet('module install dsestero-download_uncompress'))
+    on(agent, "/opt/puppetlabs/puppet/bin/puppet module install dsestero-download_uncompress")
```
Comment (cowofevil) [inline tests/integration/pre-suite/01_websphere_module_install.rb]: You used double quotes when single quotes would suffice.

File: tests/integration/tests/01_install_ibm_installation_manager_from_http.rb [hunk 4]
```diff
@@ -12,10 +12,16 @@
   end
 end
 
+if agent['platform'] =~ /aix/
+  source = 'http://int-resources.ops.puppetlabs.net/QA_resources/ibm_websphere/agent.installer.aix.gtk.ppc_1.8.4000.20151125_0201.zip'
+elsif agent['platform'] =~ /linux/
+  source = 'http://int-resources.ops.puppetlabs.net/QA_resources/ibm_websphere/agent.installer.linux.gtk.x86_64_1.8.3000.20150606_0047.zip'
+end
```
Comment (cowofevil) [inline tests/integration/tests/01_install_ibm_installation_manager_from_http.rb]: This block of code is repeated in different test cases. You should move this into a helper method to reduce code maintenance in the future.

File: tests/integration/tests/03_install_ibm_installation_manager_to_custom_location.rb [hunk 5]
```diff
@@ -14,10 +14,15 @@
   end
 end
 
+if agent['platform'] =~ /aix/
+  source = 'http://int-resources.ops.puppetlabs.net/QA_resources/ibm_websphere/agent.installer.aix.gtk.ppc_1.8.4000.20151125_0201.zip'
+elsif agent['platform'] =~ /linux/
+  source = 'http://int-resources.ops.puppetlabs.net/QA_resources/ibm_websphere/agent.installer.linux.gtk.x86_64_1.8.3000.20150606_0047.zip'
+end
```
Comment (cowofevil) [inline tests/integration/tests/03_install_ibm_installation_manager_to_custom_location.rb]: This block of code is repeated in different test cases. You should move this into a helper method to reduce code maintenance in the future.

File: tests/integration/pre-suite/01_websphere_module_install.rb [hunk 6]
```diff
@@ -1,29 +1,22 @@
...
 
-#Currently the puppetlabs-ibm_installation_manager is not ready for testing, so using joshbeard-ibm_installation_manager
-# instead.
-on(master, puppet('module install joshbeard-ibm_installation_manager'))
-
-on(master, puppet('module install nanliu-staging -v 1.0.3'))
+step 'install IBM_Installation_Manager dependencies:'
+on(master, puppet('module install puppet-archive'))
 on(master, puppet('module install puppetlabs-stdlib'))
 on(master, puppet('module install puppetlabs-concat'))
```
Comment (cowofevil) [inline tests/integration/pre-suite/01_websphere_module_install.rb]: Suggestion on a different way to write this in Ruby:

```
%w(puppet-archive puppetlabs-stdlib puppetlabs-concat).each do |dep|
  on(agent, puppet("module install #{dep}"))
end
```

File: tests/integration/pre-suite/01_websphere_module_install.rb (hunk 1 above)
Comment (phongdly) [inline tests/integration/pre-suite/01_websphere_module_install.rb]: fixed

File: tests/test_run_scripts/integration_tests_aix.sh (hunk 2 above)
Comment (phongdly) [inline tests/test_run_scripts/integration_tests_aix.sh]: fixed, deleted the lines

File: tests/integration/pre-suite/01_websphere_module_install.rb (hunk 3 above)
Comment (phongdly) [inline tests/integration/pre-suite/01_websphere_module_install.rb]: fixed

File: tests/integration/tests/01_install_ibm_installation_manager_from_http.rb (hunk 4 above)
Comment (phongdly) [inline tests/integration/tests/01_install_ibm_installation_manager_from_http.rb]: fixed, added method 'get_source_link' in the library help file

File: tests/integration/tests/03_install_ibm_installation_manager_to_custom_location.rb (hunk 5 above)
Comment (phongdly) [inline tests/integration/tests/03_install_ibm_installation_manager_to_custom_location.rb]: fixed, added method 'get_source_link' in the library help file

File: tests/integration/pre-suite/01_websphere_module_install.rb (hunk 6 above)
Comment (phongdly) [inline tests/integration/pre-suite/01_websphere_module_install.rb]: updated with the suggestion

**Comment by phongdly** [comment]
@cowofevil : updated!

File: tests/test_run_scripts/integration_tests_aix.sh [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+BASENAME_CMD="basename ${SCRIPT_PATH}"
+SCRIPT_BASE_PATH=`eval ${BASENAME_CMD}`
+
+if [ $SCRIPT_BASE_PATH = "test_run_scripts" ]; then
+  cd ../../
+fi
+
+export pe_dist_dir="http://pe-releases.puppetlabs.lan/2015.3.3"
+export GEM_SOURCE=http://rubygems.delivery.puppetlabs.net
+export BEAKER_VERSION=2.39.0
```
Comment (cowofevil) [inline tests/test_run_scripts/integration_tests_aix.sh]: What is this envar used for?

File: tests/lib/im_helper.rb [hunk 8]
```diff
@@ -70,17 +78,43 @@ def download_and_uncompress(host, urllink, compressed_file, uncompress_to)
...
+#
+# get_resource_link
+def get_source_link(host)
+  source = nil
+  if (host['platform'] =~ /aix/)
+    source = 'http://int-resources.ops.puppetlabs.net/QA_resources/ibm_websphere/agent.installer.aix.gtk.ppc_1.8.4000.20151125_0201.zip'
+  elsif (host['platform'] =~ /linux/)
+    source = 'http://int-resources.ops.puppetlabs.net/QA_resources/ibm_websphere/agent.installer.linux.gtk.x86_64_1.8.3000.20150606_0047.zip'
+  end
+  return source
```
Comment (cowofevil) [inline tests/lib/im_helper.rb]: You can save yourself some lines of code without reducing readability by doing this:

```
if (host['platform'] =~ /aix/)
+    return 'http://int-resources.ops.puppetlabs.net/QA_resources/ibm_websphere/agent.installer.aix.gtk.ppc_1.8.4000.20151125_0201.zip'
+  elsif (host['platform'] =~ /linux/)
+    return 'http://int-resources.ops.puppetlabs.net/QA_resources/ibm_websphere/agent.installer.linux.gtk.x86_64_1.8.3000.20150606_0047.zip'
+  end
```

File: tests/integration/tests/install_ibm_installation_manager_from_extracted_source.rb [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+# Teardown
+teardown do
+  confine_block(:except, :roles => %w{master dashboard database}) do
+    agents.each do |agent|
+      on(agent, "/var/ibm/InstallationManager/uninstall/uninstallc",
+         :acceptable_exit_codes => [0,127]) do |result|
+        assert_no_match(/Error/, result.stderr, 'Failed to uninstall IBM Installation Manager')
+      end
+      on(agent, "rm -rf /opt/IBM")
+      on(agent, "rm -rf #{extracted_source}/*")
```
Comment (cowofevil) [inline tests/integration/tests/install_ibm_installation_manager_from_extracted_source.rb]: How often do you need to uninstall "IBM Installation Manager"? If you do it frequently then I would suggest putting this into your helper library.

File: tests/test_run_scripts/integration_tests_aix.sh (hunk 7 above)
Comment (phongdly) [inline tests/test_run_scripts/integration_tests_aix.sh]: actually this line is not needed, already defined in Gemfile, will delete in the next push

File: tests/lib/im_helper.rb (hunk 8 above)
Comment (phongdly) [inline tests/lib/im_helper.rb]: Will update in the next push

File: tests/integration/tests/install_ibm_installation_manager_from_extracted_source.rb (hunk 9 above)
Comment (phongdly) [inline tests/integration/tests/install_ibm_installation_manager_from_extracted_source.rb]: Good idea, this test to install and uninstall IM, so the uninstall is very often, will update in the next push

**Comment by phongdly** [comment]
@cowofevil : updated, tested, and pushed.
Note: I clean the AIX box so you can run the test anytime.

**Pull request merged**